### DISCUSSION TITLE
Citations in MS Teams

### DISCUSSION THREAD

**Discussion by jdmevo123** [microsoft/teams-sdk#1809] [Q&A]
I currently have a teams bot with the teams-ai library. I cannot get citations to work if no url is present. I am following the sayCommand extract of citations, which is working, but in teams it doesnt provide the citations unless the URL is present, but in my case I do not have URL.

Here is my code for citations:

`app.ai.action<PredictedSayCommand>(AI.SayCommandActionName, async (context, state, data, action) => {
    let content = data.response.content || '';
    const isTeamsChannel = context.activity.channelId === Channels.Msteams;

    if (isTeamsChannel) {
        content = content.split('\n').join('<br>');
    }
    let citations: ClientCitation[] | undefined = undefined;

    if (data.response.context && data.response.context.citations.length > 0) {
        citations = data.response.context!.citations.map((citation, i) => {
            return {
                '@type': 'Claim',
                position: `${i + 1}`,
                appearance: {
                    '@type': 'DigitalDocument',
                    name: citation.filepath,
                    abstract: snippet(citation.content, 500),
                    usageInfo: {
                        "@type": "CreativeWork",
                        name: "Confidential",
                        description: "The information identified in this message is Privileged and Confidential and should not be shared with unauthorized personnel.",
                      },
                }
            } as ClientCitation;
        });
    }
    const contentText = data.response?.content?.replace(/\[doc(\d+)\]/g, '[$1]');

    const referencedCitations = citations ? getUsedCitations(content, citations) : undefined;

    await context.sendActivity({
        type: ActivityTypes.Message,
        text: contentText,
        ...(isTeamsChannel ? { channelData: { feedbackLoopEnabled: true } } : {}),
        entities: [
            {
                type: 'https://schema.org/Message',
                '@type': 'Message',
                '@context': 'https://schema.org',
                '@id': '',
                usageInfo: {
                            "@type": "CreativeWork",
                            name: "Confidential",
                            description: "The information identified in this message is Privileged and Confidential and should not be shared with unauthorized personnel.",
                          },
                additionalType: ['AIGeneratedContent'],
                ...(referencedCitations ? { citation: referencedCitations } : {})
            }
        ] as Entities[]
    return "";
});`

In teams I just get the text with the citation [1]. However the citation is not clickable or no chevron with a list of the documents.

**Comment by corinagum** [comment]
Based off what you've written, there has probably been a change in the UX for Teams platform so I will bring this up internally with the UI team. Thanks for the heads up.

**Reply by corinagum** [reply]
@jdmevo123 Thanks for your patience. I've spoken with the UI team and citations appear to be working as expected, with the caveat that citations in the footer are not yet supported. 
Could you share a screenshot/video of what your citations look like so I can do a visual comparison?

**Reply by corinagum** [reply]
@jdmevo123 If you could also provide the activity payload of what is being sent to the bot when you expect to receive citations back, that would be great. Otherwise, could you put a quick repro into a repository so we can take a look?
cc @premkaretimsft

**Reply by jdmevo123** [reply]
@corinagum @premkaretimsft

The payload is in the discussion, it’s using the sayCommand response as per the samples, this should construct the citations.

Are you saying this method no longer works and we should we should send this as an action?

**Reply by jdmevo123** [reply]
This is what the citations come through as, the linked ones are the URL’s and if you want to display the document without a url, the citation is not a link.
![IMG_9465](https://github.com/microsoft/teams-ai/assets/6053506/2a7fb6db-7594-45d9-a0bb-6542bdc6e7f4)
![IMG_9466](https://github.com/microsoft/teams-ai/assets/6053506/db78f482-d9a5-49dd-bb81-6cade94f85d5)

**Reply by jdmevo123** [reply]
this is the documentation on teams-ai and citations.

https://github.com/microsoft/teams-ai/blob/main/getting-started/CONCEPTS/POWERED-BY-AI.md

**Reply by corinagum** [reply]
Can you confirm what RAG service you are using? Azure search with On Your Data?

Also, if you are not using developer preview, you will need to enable that as citations UI is not publicly available yet. The estimated release for public use (desktop) is within the next few weeks :)

Please disregard what I said about citations in footer - since you provided the screencaps I can see it is unrelated to your issue.

**Reply by jdmevo123** [reply]
@corinagum, Im using bring your own data with MongoDB

I am using teams preview i believe, please see attached screen shot. However Citations are still not working.
![Screenshot 2024-07-11 091542](https://github.com/microsoft/teams-ai/assets/6053506/4b205f6a-6bba-478f-a90a-aaca434c7e60)

**Reply by premkaretimsft** [reply]
@jdmevo123 , could you log the raw payload of the outgoing network request and response of
the `sendActivity` for both cases i.e. with URL and without URL and share that data?

**Comment by ffarukkoc** [comment]
Hello,
I tried to run the code on "https://github.com/microsoft/teams-ai/issues/2149" issue but I couldn't succeed. It gives "[Error] - code:Core.DecryptionError, message: Cipher text is broken"error.

Alternatively, I added data source to my project as following image.

![image](https://github.com/user-attachments/assets/23066b57-24d4-4c73-82f6-41aea092fb64)


It also gave error :( "The chat completion API returned an error
status of None: Error code: 400 - {'error': {'message': 'Unrecognized request argument supplied: data_sources', 'type': 'invalid_request_error', 'param': None, 'code': None}}"

I tried to follow the code and I noticed that because of data source part it gives error.

![image](https://github.com/user-attachments/assets/11e022b4-b340-4a6b-ae6f-657173d52cdf)


Do you have any suggestion to solve this problem?

**Reply by ffarukkoc** [reply]
Hello,
I found the solution. Main problem is on data source definition. First of all, you have to check did your index saved as data source. You can do it from Azure AI Studio ([see example](https://learn.microsoft.com/en-us/azure/ai-services/openai/use-your-data-quickstart?tabs=command-line%2Cjavascript-keyless%2Ctypescript-keyless%2Cpython-new&pivots=programming-language-studio#add-your-data-using-azure-openai-studio)). Then you have to add data source part to config.json file. Whlie doing that don't forgot to delete extra information on augmentation part.

**Comment by heyitsaamir** [comment]
We're clearing discussions and starting fresh. This discussion has not been acted upon. Please reopen the discussion, or start a new one if this question is still important. Thank you for your understanding.
